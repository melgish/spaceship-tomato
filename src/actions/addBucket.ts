'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddBucket, addBucketSchema } from '@/schemas/addBucket.schema'

/**
 * Add server action to a new bucket to the database.
 * @see ../schemas/addBucket.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addBucket(payload: AddBucket) {
  try {
    const data = await addBucketSchema.parseAsync(payload)
    const created = await prisma.bucket.create({
      data: { ...data, plantId: data.plantId || null },
    })
    revalidatePath('/buckets')
    return { ok: true, message: 'Bucket created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create bucket' } as const
  }
}
