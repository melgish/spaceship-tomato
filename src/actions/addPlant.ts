'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddPlant, addPlantSchema } from '@/schemas/addPlant.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addPlant.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addPlant(payload: AddPlant) {
  try {
    const data = await addPlantSchema.parseAsync(payload)
    const created = await prisma.plant.create({ data })
    revalidatePath('/plants')
    return { ok: true, message: 'Plant created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create plant' } as const
  }
}
