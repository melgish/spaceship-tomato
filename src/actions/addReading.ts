'use server'
import { revalidatePath } from 'next/cache'
import { prisma } from '@/lib/prisma'
import { type AddReading, addReadingSchema } from '@/schemas/addReading.schema'

/**
 * Add server action to a new plant to the database.
 * @see ../schemas/addReading.schema.ts for validation rules.
 * @param payload The request payload
 * @returns Ok result when successful, otherwise an Err result.
 */
export async function addReading(payload: AddReading) {
  try {
    const data = await addReadingSchema.parseAsync(payload)
    const created = await prisma.reading.create({ data })
    revalidatePath('/readings')
    return { ok: true, message: 'Reading created', data: created } as const
  } catch (error: any) {
    return { ok: false, message: 'Failed to create reading' } as const
  }
}
