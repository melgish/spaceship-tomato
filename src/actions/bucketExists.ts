'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that a bucket exists in the database.
 * @param id Id of bucket to search for.
 * @returns true if bucket exists, otherwise false.
 */
export async function bucketExists(id: number = 0) {
  return id > 0 && (await prisma.bucket.count({ where: { id } })) > 0
}
