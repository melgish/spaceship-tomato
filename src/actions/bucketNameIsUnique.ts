'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to validate that the name of a bucket is unique.
 * @param name Name of the bucket to test.
 * @param id Optional id of bucket to exclude when editing an existing bucket.
 * @returns true if the name exists on another bucket, otherwise false.
 */
export async function bucketNameIsUnique(name: string, id: number = 0) {
  const where = { name, id: (id > 0 && { not: id }) || undefined }
  return (await prisma.bucket.count({ where })) === 0
}
