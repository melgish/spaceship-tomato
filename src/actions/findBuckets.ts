'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of buckets from the database.
 * @returns An array of buckets. The returned data will include plant info if
 * the bucket contains a plant.
 */
export async function findBuckets() {
  return await prisma.bucket.findMany({
    include: { plant: true },
    orderBy: [{ name: 'asc' }],
  })
}

export type Buckets = Awaited<ReturnType<typeof findBuckets>>
export type Bucket = Buckets[number]
