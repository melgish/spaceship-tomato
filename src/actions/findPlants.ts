'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all of plants from the database.
 * @returns An array of plants.
 */
export async function findPlants() {
  return await prisma.plant.findMany({
    orderBy: [{ name: 'asc' }],
  })
}

export type Plants = Awaited<ReturnType<typeof findPlants>>
export type Plant = Plants[number]
