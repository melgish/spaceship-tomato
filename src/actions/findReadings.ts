'use server'
import { prisma } from '@/lib/prisma'

/**
 * A server action to fetch all readings from the database.
 * @returns An array of readings.  Readings will include bucket and plant info.
 */
export async function findReadings() {
  return await prisma.reading.findMany({
    orderBy: [{ createdAt: 'desc' }],
    include: {
      bucket: {
        include: {
          plant: true,
        },
      },
    },
  })
}

export type Readings = Awaited<ReturnType<typeof findReadings>>
export type Reading = Readings[number]
