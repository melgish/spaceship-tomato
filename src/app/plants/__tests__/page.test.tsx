import { type Mock, beforeEach, expect, it, vi, describe } from 'vitest'
import { render } from '@testing-library/react'
import { findBuckets } from '@/actions/findBuckets'
import { findPlants } from '@/actions/findPlants'
import BucketsPage from '@/app/buckets/page'

vi.mock('@/actions/findBuckets')
vi.mock('@/actions/findPlants')

beforeEach(() => {
  ;(findBuckets as Mock).mockResolvedValue([])
  ;(findPlants as Mock).mockResolvedValue([])
})

describe('when there is no data', () => {
  it('should render', async () => {
    const { container } = render(await BucketsPage())
    expect(container).toMatchSnapshot()
  })
})
