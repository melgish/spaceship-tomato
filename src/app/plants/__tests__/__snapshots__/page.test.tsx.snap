// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when there is no data > should render 1`] = `
<div>
  <section>
    <header>
      <div>
        Buckets
      </div>
    </header>
    <form
      class="flex gap-x-3"
      novalidate=""
    >
      <div
        class="required field"
      >
        <label
          for=":r0:-name"
        >
          Name
        </label>
        <input
          autocomplete="off"
          id=":r0:-name"
          maxlength="24"
          name="name"
          placeholder="Ex. #1"
          required=""
          size="24"
          type="text"
        />
      </div>
      <div
        class="field"
      >
        <label
          for=":r0:-plantId"
        >
          Plant
        </label>
        <select
          class="py-1.5"
          id=":r0:-plantId"
          name="plantId"
          style="width: 32ch;"
        >
          <option
            value="0"
          >
            None
          </option>
        </select>
      </div>
      <div
        class="flex items-center gap-x-2"
      >
        <button
          class="emerald"
          type="submit"
        >
          Submit
        </button>
        <button
          class="slate"
          type="button"
        >
          Clear
        </button>
      </div>
    </form>
    <div
      class="mt-2"
    />
  </section>
</div>
`;
